internal/client/conn.go

```diff
@@ -0,0 +1,305 @@
+)
+
+var noDeadline time.Time
+
+type inboundData struct {
+	data []byte
+	from net.Addr
+}
+
+// UDPConnObserver is an interface to UDPConn observer
+type UDPConnObserver interface {
+	TURNServerAddr() net.Addr
+	Username() stun.Username
+	Realm() stun.Realm
+	WriteTo(data []byte, to net.Addr) (int, error)
+	PerformTransaction(msg *stun.Message, to net.Addr, dontWait bool) (TransactionResult, error)
+}
+
+// UDPConnConfig is a set of configuration params use by NewUDPConn
+type UDPConnConfig struct {
+	Observer  UDPConnObserver
+	Relayed   turn.RelayedAddress
+	Integrity stun.MessageIntegrity
+	Nonce     stun.Nonce
+	Lifetime  time.Duration
+	Log       logging.LeveledLogger
+}
+
+// UDPConn is the implementation of the Conn and PacketConn interfaces for UDP network connections.
+// comatible with net.PacketConn and net.Conn
+type UDPConn struct {
+	obs        UDPConnObserver
+	relayed    turn.RelayedAddress
+	permMap    *permissionMap
+	channelMap map[string]*channel
+	bindingMap map[turn.ChannelNumber]*channel
+	minBound   turn.ChannelNumber
+	integrity  stun.MessageIntegrity
+	nonce      stun.Nonce
+	lifetime   time.Duration
+	readCh     chan *inboundData
+	closeCh    chan struct{}
+	readTimer  *time.Timer
+	log        logging.LeveledLogger
+}
+
+// NewUDPConn creates a new instance of UDPConn
+func NewUDPConn(config *UDPConnConfig) *UDPConn {
+	conn := &UDPConn{
+		obs:        config.Observer,
+		relayed:    config.Relayed,
+		permMap:    newPermissionMap(),
+		channelMap: map[string]*channel{},
+		bindingMap: map[turn.ChannelNumber]*channel{},
+		minBound:   turn.MinChannelNumber,
+		integrity:  config.Integrity,
+		nonce:      config.Nonce,
+		lifetime:   config.Lifetime,
+		readCh:     make(chan *inboundData, maxReadQueueSize),
+		closeCh:    make(chan struct{}),
+		readTimer:  time.NewTimer(time.Duration(math.MaxInt64)),
+		log:        config.Log,
+	}
+
+	conn.log.Debugf("lifetime is %d seconds", int(conn.lifetime.Seconds()))
+
+	return conn
+}
+
+// ReadFrom reads a packet from the connection,
+// copying the payload into p. It returns the number of
+// bytes copied into p and the return address that
+// was on the packet.
+// It returns the number of bytes read (0 <= n <= len(p))
+// and any error encountered. Callers should always process
+// the n > 0 bytes returned before considering the error err.
+// ReadFrom can be made to time out and return
+// an Error with Timeout() == true after a fixed time limit;
+// see SetDeadline and SetReadDeadline.
+func (c *UDPConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+loop:
+	for {
+		select {
+		case ibData := <-c.readCh:
+			n := copy(p, ibData.data)
+			if n < len(ibData.data) {
+				return 0, nil, io.ErrShortBuffer
+			}
+			return n, ibData.from, nil
+
+		case <-c.readTimer.C:
+			return 0, nil, &net.OpError{
+				Op:   "read",
+				Net:  c.LocalAddr().Network(),
+				Addr: c.LocalAddr(),
+				Err:  newTimeoutError("i/o timeout"),
+			}
+
+		case <-c.closeCh:
```

Comment:
Can we return an error here, are you worried there might be more data to read?  If so I would vote for just another read call (just to be explicit). I had to read this a few times to make sure I didn't miss anything!

Reply:
I can't remember why I added the closeCh... (I copied this code from vnet) Let me review.

Reply:
closeCh was necessary because it was not safe to close readCh, if I remember correctly. But, there's anoteher problem. User may call ReadFrom() after the conn is closed. At that point, it may happen to read data again (if it is the buffer), which is not the expected behavior.  For that, I have added a atomic flag `closed` (I introduced `AtomicBool`) in order to return error when that happens.

Reply:
Nice! Didn't think of that.

Do we need the `break loop`? The `for` loop should run again and bail I believe?

Reply:
Oh true. Will remove the break/loop.